### PR TITLE
Created a cart controller

### PR DESCRIPTION
**Pull request by mikkoh**
This PR create a `CartController` that the `ThemeControllerBase` has an instance to. Eventually this could lead to all `ThemeControllers` using the same `CartController` instance.

### CODE REVIEW

File: Assets/Shopify/UIToolkit/Theme Controllers/CartController.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+        [System.Serializable]
+        public class PurchaseFailedEvent : UnityEvent<ShopifyError> {}
+
+        private Cart Cart;
+        public UnityEvent OnPurchaseStarted = new UnityEvent();
+        public QuantityChangeEvent OnQuantityChange = new QuantityChangeEvent();
+        public UnityEvent OnPurchaseComplete = new UnityEvent();
+        public UnityEvent OnPurhcaseCancelled = new UnityEvent();
+        public PurchaseFailedEvent OnPurhcaseFailed = new PurchaseFailedEvent();
```
Comment (sleroux) [inline Assets/Shopify/UIToolkit/Theme Controllers/CartController.cs]: Nit: Typo `OnPurhcaseFailed` -> `OnPurchaseFailed`

Comment (sleroux) [reply Assets/Shopify/UIToolkit/Theme Controllers/CartController.cs]: Same with Cancelled

File: Assets/Shopify/UIToolkit/Theme Controllers/CartController.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            SetCart(cart);
+        }
+
+        public void SetCart(Cart cart) {
+            Cart = cart;
+        }
+
+        public CartLineItems LineItems {
+            get {
+                return Cart.LineItems;
```
Comment (mikkoh) [inline Assets/Shopify/UIToolkit/Theme Controllers/CartController.cs]: Should be noted that currently if you go and modify quantities directly on line items we're not reactive to this. Will need to create a second PR which adds eventing to the `Cart` that propagates line item changes.

File: Assets/Shopify/UIToolkit/Theme Controllers/ThemeControllerBase.cs [hunk 3]
```diff
@@ -65,13 +65,36 @@ public abstract class ThemeControllerBase : MonoBehaviour, IShopCredentials {
...
+            get {
+                if (_cachedClient == null) {
+                    SetupClientAndCart();
+                }
+
+                return _cachedCart;
+            }
+        }
+
+        private void SetupClientAndCart() {
```
Comment (JakeCataford) [inline Assets/Shopify/UIToolkit/Theme Controllers/ThemeControllerBase.cs]: Im weary about methods that have a "SomethingAndSomething", maybe its better to call this something along the lines of "InitializeDependencies" or something more generic?

File: Assets/Shopify/UIToolkit/Test/Unit/Editor/TestThemeControllerBase.cs [hunk 4]
```diff
@@ -193,10 +193,10 @@ public class MockThemeController : ThemeControllerBase {
                 yield return null;
             }
 
-            _themeController.UpdateVariantInCart(_variants.First(), 1);
-            _themeController.UpdateVariantInCart(_variants[1], 2);
+            _themeController.Cart.UpdateVariant(_variants.First(), 1);
+            _themeController.Cart.UpdateVariant(_variants[1], 2);
 
-            _themeController.ClearCart();
+            _themeController.Cart.ClearCart();
```
Comment (JakeCataford) [inline Assets/Shopify/UIToolkit/Test/Unit/Editor/TestThemeControllerBase.cs]: One of the advantages of breaking apart the cart is that we can test its functionality as a separate unit... how hard would it be to pull these out into a TestCartController or something similar?

File: Assets/Shopify/UIToolkit/Theme Controllers/ThemeControllerBase.cs [hunk 5]
```diff
@@ -65,13 +65,36 @@ public abstract class ThemeControllerBase : MonoBehaviour, IShopCredentials {
...
+            _cachedClient = new ShopifyClient(LoaderProvider.GetLoader(AccessToken, ShopDomain));
+            _cachedCart = new CartController(_cachedClient.Cart());
+            
+            _cachedCart.OnPurchaseStarted.AddListener(Theme.OnPurchaseStarted);
+            _cachedCart.OnPurhcaseCancelled.AddListener(Theme.OnPurchaseCancelled);
+            _cachedCart.OnPurchaseComplete.AddListener(Theme.OnPurchaseCompleted);
+            _cachedCart.OnPurhcaseFailed.AddListener(Theme.OnPurchaseFailed);
```
Comment (JakeCataford) [inline Assets/Shopify/UIToolkit/Theme Controllers/ThemeControllerBase.cs]: typo here

File: Assets/Shopify/UIToolkit/Theme Controllers/CartController.cs [hunk 6]
```diff
@@ -0,0 +1,122 @@
+namespace Shopify.UIToolkit {
+    using UnityEngine.Events;
+    using Shopify.Unity;
+    using Shopify.Unity.SDK;
+    using System.Linq;
+
+    public class CartController {
```
Comment (JakeCataford) [inline Assets/Shopify/UIToolkit/Theme Controllers/CartController.cs]: I think you want this to be serializable as well

File: Assets/Shopify/UIToolkit/Theme Controllers/CartController.cs (hunk 1 above)
Comment (JakeCataford) [inline Assets/Shopify/UIToolkit/Theme Controllers/CartController.cs]: We will need to expose these in an editor in the future, but its probably best to do so when doing the rest of the events

File: Assets/Shopify/UIToolkit/Theme Controllers/ThemeControllerBase.cs [hunk 7]
```diff
@@ -65,13 +65,36 @@ public abstract class ThemeControllerBase : MonoBehaviour, IShopCredentials {
...
                 if (_cachedClient == null) {
-                    _cachedClient = new ShopifyClient(LoaderProvider.GetLoader(AccessToken, ShopDomain));
+                    SetupClientAndCart();
                 }
                 return _cachedClient;
             }
         }
         private ShopifyClient _cachedClient;
 
+        public CartController Cart {
```
Comment (JakeCataford) [inline Assets/Shopify/UIToolkit/Theme Controllers/ThemeControllerBase.cs]: Dont think its nescesary right now but its neat that in the future we could make CartController implement ICartController and allow dependencies to be injected here

File: Assets/Shopify/UIToolkit/Test/Unit/Editor/TestThemeControllerBase.cs (hunk 4 above)
Comment (mikkoh) [reply Assets/Shopify/UIToolkit/Test/Unit/Editor/TestThemeControllerBase.cs]: Shouldn't be too tough and that's a good point.

File: Assets/Shopify/UIToolkit/Theme Controllers/CartController.cs (hunk 6 above)
Comment (mikkoh) [reply Assets/Shopify/UIToolkit/Theme Controllers/CartController.cs]: Will nested classes/instances be rendered in the editor? So like the controller will render but I'm not sure if the `UnityEvents` for the cart will currently be useable.

File: Assets/Shopify/UIToolkit/Theme Controllers/ThemeControllerBase.cs (hunk 7 above)
Comment (mikkoh) [reply Assets/Shopify/UIToolkit/Theme Controllers/ThemeControllerBase.cs]: Yeah. Agreed. Also we should make a `IThemeController`

File: Assets/Shopify/UIToolkit/Theme Controllers/ThemeControllerBase.cs (hunk 3 above)
Comment (mikkoh) [reply Assets/Shopify/UIToolkit/Theme Controllers/ThemeControllerBase.cs]: I worry that `InitializeDependencies` is too generic

**Pull request merged**